File: backend/app/services/content_ideas.py
import json
from typing import Any

from sqlalchemy import case, delete, exists, func, or_
from sqlmodel import Session, select

from app.models.content_idea import (
    ContentIdea,
    ContentIdeaSource,
    ContentIdeaSort,
    ContentIdeaStatus,
    ContentPlatform,
    ContentPriority,
)
from app.models.content_idea_broll import ContentIdeaBroll
from app.models.content_idea_reference import ContentIdeaReference
from app.models.user import utc_now
from app.models.project import Project
from app.models.saved_broll import SavedBroll
from app.models.saved_reference import SavedReference
from app.schemas.content_idea import (
    ContentIdeaCreate,
    ContentIdeaConversionCreate,
    ContentIdeaRead,
    ContentIdeaPriorityCounts,
    ContentIdeaPlatformCounts,
    ContentIdeaSourceCounts,
    ContentIdeaStatusCounts,
    ContentIdeaSummaryRead,
    ContentIdeaUpdate,
    normalize_tags as _normalize_tags,
)
from app.schemas.project import ProjectRead


class ContentIdeaNotFoundError(Exception):
    pass


class ContentIdeaConversionConflictError(Exception):
    def __init__(self, reason: str) -> None:
        super().__init__(reason)
        self.reason = reason


def normalize_tags(tags: Any) -> Any:
    return _normalize_tags(tags)


def serialize_tags(tags: list[str]) -> str | None:
    if not tags:
        return None
    return json.dumps(tags, ensure_ascii=False)


def deserialize_tags(value: str | None) -> list[str]:
    if not value:
        return []
    try:
        parsed = json.loads(value)
    except (TypeError, json.JSONDecodeError):
        return []
    if not isinstance(parsed, list):
        return []
    return [tag for tag in parsed if isinstance(tag, str)]


def to_content_idea_read(idea: ContentIdea) -> ContentIdeaRead:
    assert idea.id is not None
    return ContentIdeaRead(
        id=idea.id,
        user_id=idea.user_id,
        title=idea.title,
        description=idea.description,
        platform=ContentPlatform(idea.platform),
        status=ContentIdeaStatus(idea.status),
        priority=ContentPriority(idea.priority),
        tags=deserialize_tags(idea.tags),
        target_audience=idea.target_audience,
        content_format=idea.content_format,
        source=ContentIdeaSource(idea.source),
        converted_project_id=idea.converted_project_id,
        created_at=idea.created_at,
        updated_at=idea.updated_at,
    )


def _normalize_nullable_text_fields(data: dict[str, Any]) -> None:
    for field_name in ("description", "target_audience", "content_format"):
        if field_name in data and data[field_name] == "":
            data[field_name] = None


def list_content_ideas(
    session: Session,
    user_id: int,
    *,
    status: ContentIdeaStatus | None = None,
    platform: ContentPlatform | None = None,
    priority: ContentPriority | None = None,
    source: ContentIdeaSource | None = None,
    search: str | None = None,
    sort: ContentIdeaSort = ContentIdeaSort.CREATED_DESC,
) -> list[ContentIdeaRead]:
    statement = select(ContentIdea).where(ContentIdea.user_id == user_id)
    if status is not None:
        statement = statement.where(ContentIdea.status == status)
    if platform is not None:
        statement = statement.where(ContentIdea.platform == platform)
    if priority is not None:
        statement = statement.where(ContentIdea.priority == priority)
    if source is not None:
        statement = statement.where(ContentIdea.source == source)
    if search and (term := search.strip()):
        escaped = term.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
        pattern = f"%{escaped}%"
        statement = statement.where(
            or_(
                ContentIdea.title.ilike(pattern, escape="\\"),
                ContentIdea.description.ilike(pattern, escape="\\"),
            )
        )

    order_by = {
        ContentIdeaSort.CREATED_DESC: (ContentIdea.created_at.desc(), ContentIdea.id.desc()),
        ContentIdeaSort.CREATED_ASC: (ContentIdea.created_at.asc(), ContentIdea.id.asc()),
        ContentIdeaSort.UPDATED_DESC: (ContentIdea.updated_at.desc(), ContentIdea.id.desc()),
        ContentIdeaSort.UPDATED_ASC: (ContentIdea.updated_at.asc(), ContentIdea.id.asc()),
        ContentIdeaSort.TITLE_ASC: (ContentIdea.title.asc(), ContentIdea.id.asc()),
        ContentIdeaSort.TITLE_DESC: (ContentIdea.title.desc(), ContentIdea.id.desc()),
    }
    if sort in (ContentIdeaSort.PRIORITY_DESC, ContentIdeaSort.PRIORITY_ASC):
        priority_rank = case(
            (ContentIdea.priority == ContentPriority.LOW, 1),
            (ContentIdea.priority == ContentPriority.MEDIUM, 2),
            (ContentIdea.priority == ContentPriority.HIGH, 3),
            else_=0,
        )
        priority_order = (
            priority_rank.desc()
            if sort == ContentIdeaSort.PRIORITY_DESC
            else priority_rank.asc()
        )
        statement = statement.order_by(
            priority_order,
            ContentIdea.updated_at.desc(),
            ContentIdea.id.desc(),
        )
    else:
        statement = statement.order_by(*order_by[sort])
    return [to_content_idea_read(idea) for idea in session.exec(statement).all()]


def _enum_counts(
    session: Session,
    user_id: int,
    column: Any,
) -> dict[str, int]:
    statement = (
        select(column, func.count(ContentIdea.id))
        .where(ContentIdea.user_id == user_id)
        .group_by(column)
    )
    return {str(value): count for value, count in session.exec(statement).all()}


def get_content_idea_summary(
    session: Session,
    user_id: int,
) -> ContentIdeaSummaryRead:
    total, latest_updated_at = session.exec(
        select(func.count(ContentIdea.id), func.max(ContentIdea.updated_at)).where(
            ContentIdea.user_id == user_id
        )
    ).one()
    by_status = _enum_counts(session, user_id, ContentIdea.status)
    by_priority = _enum_counts(session, user_id, ContentIdea.priority)
    by_platform = _enum_counts(session, user_id, ContentIdea.platform)
    by_source = _enum_counts(session, user_id, ContentIdea.source)

    reference_exists = exists().where(
        ContentIdeaReference.content_idea_id == ContentIdea.id
    )
    broll_exists = exists().where(ContentIdeaBroll.content_idea_id == ContentIdea.id)

    def count_where(*conditions: Any) -> int:
        return session.exec(
            select(func.count(ContentIdea.id)).where(
                ContentIdea.user_id == user_id,
                *conditions,
            )
        ).one()

    return ContentIdeaSummaryRead(
        total=total,
        by_status=ContentIdeaStatusCounts(**by_status),
        by_priority=ContentIdeaPriorityCounts(**by_priority),
        by_platform=ContentIdeaPlatformCounts(**by_platform),
        by_source=ContentIdeaSourceCounts(**by_source),
        converted_count=count_where(ContentIdea.converted_project_id.is_not(None)),
        ready_count=by_status.get(ContentIdeaStatus.READY, 0),
        active_count=sum(
            by_status.get(status, 0)
            for status in (
                ContentIdeaStatus.IDEA,
                ContentIdeaStatus.RESEARCHING,
                ContentIdeaStatus.READY,
            )
        ),
        with_reference_count=count_where(reference_exists),
        with_broll_count=count_where(broll_exists),
        with_any_media_count=count_where(or_(reference_exists, broll_exists)),
        latest_updated_at=latest_updated_at,
    )


def create_content_idea(
    session: Session,
    user_id: int,
    idea_create: ContentIdeaCreate,
) -> ContentIdeaRead:
    return create_content_idea_with_source(
        session,
        user_id,
        idea_create,
        source=ContentIdeaSource.MANUAL,
    )


def create_content_idea_with_source(
    session: Session,
    user_id: int,
    idea_create: ContentIdeaCreate,
    *,
    source: ContentIdeaSource,
) -> ContentIdeaRead:
    data = idea_create.model_dump(exclude={"tags"})
    _normalize_nullable_text_fields(data)
    idea = ContentIdea(
        user_id=user_id,
        source=source,
        tags=serialize_tags(idea_create.tags),
        **data,
    )
    try:
        session.add(idea)
        session.commit()
        session.refresh(idea)
    except Exception:
        session.rollback()
        raise
    return to_content_idea_read(idea)


def get_content_idea(
    session: Session,
    user_id: int,
    idea_id: int,
) -> ContentIdea | None:
    statement = select(ContentIdea).where(
        ContentIdea.id == idea_id,
        ContentIdea.user_id == user_id,
    )
    return session.exec(statement).one_or_none()


def update_content_idea(
    session: Session,
    idea: ContentIdea,
    idea_update: ContentIdeaUpdate,
) -> ContentIdeaRead:
    data = idea_update.model_dump(exclude_unset=True, exclude={"tags"})
    _normalize_nullable_text_fields(data)
    idea.sqlmodel_update(data)
    if "tags" in idea_update.model_fields_set:
        idea.tags = serialize_tags(idea_update.tags or [])
    idea.updated_at = utc_now()
    session.add(idea)
    session.commit()
    session.refresh(idea)
    return to_content_idea_read(idea)


def delete_content_idea(session: Session, idea: ContentIdea) -> None:
    assert idea.id is not None
    try:
        session.exec(
            delete(ContentIdeaReference).where(
                ContentIdeaReference.content_idea_id == idea.id
            )
        )
        session.exec(
            delete(ContentIdeaBroll).where(
                ContentIdeaBroll.content_idea_id == idea.id
            )
        )
        session.delete(idea)
        session.commit()
    except Exception:
        session.rollback()
        raise


def convert_content_idea_to_project(
    session: Session,
    user_id: int,
    idea_id: int,
    data: ContentIdeaConversionCreate,
) -> tuple[ProjectRead, ContentIdeaRead]:
    idea = get_content_idea(session, user_id, idea_id)
    if idea is None:
        raise ContentIdeaNotFoundError
    if (
        idea.converted_project_id is not None
        or idea.status == ContentIdeaStatus.CONVERTED
    ):
        raise ContentIdeaConversionConflictError("converted")
    if idea.status == ContentIdeaStatus.ARCHIVED:
        raise ContentIdeaConversionConflictError("archived")

    project_data = data.model_dump()
    if project_data.get("description") == "":
        project_data["description"] = None
    project = Project(user_id=user_id, client_name=None, **project_data)

    try:
        session.add(project)
        session.flush()
        assert project.id is not None
        idea_references = list(
            session.exec(
                select(ContentIdeaReference).where(
                    ContentIdeaReference.content_idea_id == idea.id
                )
            ).all()
        )
        idea_brolls = list(
            session.exec(
                select(ContentIdeaBroll).where(ContentIdeaBroll.content_idea_id == idea.id)
            ).all()
        )
        for reference in idea_references:
            session.add(
                SavedReference(
                    project_id=project.id,
                    provider=reference.provider,
                    external_id=reference.external_id,
                    title=reference.title,
                    url=reference.url,
                    thumbnail_url=reference.thumbnail_url,
                    channel_title=reference.channel_title,
                    published_at=reference.published_at,
                    note=reference.note,
                )
            )
        for broll in idea_brolls:
            session.add(
                SavedBroll(
                    project_id=project.id,
                    provider=broll.provider,
                    external_id=broll.external_id,
                    title=broll.title,
                    url=broll.url,
                    preview_url=broll.preview_url,
                    thumbnail_url=broll.thumbnail_url,
                    creator_name=broll.creator_name,
                    duration_seconds=broll.duration_seconds,
                    width=broll.width,
                    height=broll.height,
                    note=broll.note,
                )
            )
        idea.status = ContentIdeaStatus.CONVERTED
        idea.converted_project_id = project.id
        idea.updated_at = utc_now()
        session.add(idea)
        session.commit()
        session.refresh(project)
        session.refresh(idea)
    except Exception:
        session.rollback()
        raise

    return (
        ProjectRead.model_validate(
            {
                **project.model_dump(),
                "reference_count": len(idea_references),
                "broll_count": len(idea_brolls),
                "checklist_total": 0,
                "checklist_completed": 0,
            }
        ),
        to_content_idea_read(idea),
    )

File: src/constants/contentIdeas.js
export const CONTENT_IDEA_PLATFORMS = {
  youtube: 'YouTube',
  shorts: 'YouTube Shorts',
  instagram: 'Instagram',
  tiktok: 'TikTok',
  blog: 'Blog',
  other: '기타',
}

export const CONTENT_IDEA_STATUSES = {
  idea: '아이디어',
  researching: '조사 중',
  ready: '제작 준비',
  converted: '프로젝트 전환 완료',
  archived: '보관',
}

export const CONTENT_IDEA_PRIORITIES = {
  low: '낮음',
  medium: '보통',
  high: '높음',
}

export const CONTENT_IDEA_SOURCES = {
  manual: '직접 등록',
  ai: 'AI 추천',
}

export const CONTENT_IDEA_RECOMMENDATION_TONES = {
  informative: '정보형',
  friendly: '친근한 톤',
  professional: '전문적인 톤',
  energetic: '활기찬 톤',
  humorous: '유머러스한 톤',
  inspirational: '영감을 주는 톤',
}

export const CONTENT_IDEA_RECOMMENDATION_GENRES = [
  {
    value: 'entertainment_comedy',
    label: '엔터테인먼트/코미디',
    description: '일상, 개그, 콩트, 챌린지처럼 즐거움을 주는 콘텐츠',
    topicPlaceholder: '예: 직장인 공감 상황극과 짧은 코미디',
  },
  {
    value: 'vlog_lifestyle',
    label: '브이로그/라이프스타일',
    description: '일상, 루틴, 취미와 라이프스타일을 담는 콘텐츠',
    topicPlaceholder: '예: 혼자 사는 직장인의 평일 저녁 루틴',
  },
  {
    value: 'food_cooking',
    label: '먹방/요리',
    description: '레시피, 맛집, 식문화와 먹는 즐거움을 다루는 콘텐츠',
    topicPlaceholder: '예: 자취생을 위한 10분 저녁 레시피',
  },
  {
    value: 'gaming',
    label: '게임',
    description: '플레이, 공략, 리뷰와 게임 이야기를 다루는 콘텐츠',
    topicPlaceholder: '예: 초보자를 위한 발로란트 입문 콘텐츠',
  },
  {
    value: 'beauty_fashion',
    label: '뷰티/패션',
    description: '메이크업, 스타일링, 제품과 트렌드를 소개하는 콘텐츠',
    topicPlaceholder: '예: 초보자를 위한 데일리 메이크업',
  },
  {
    value: 'knowledge_education',
    label: '지식/정보/교육',
    description: '지식, 노하우와 복잡한 정보를 쉽게 전하는 콘텐츠',
    topicPlaceholder: '예: 5분 안에 이해하는 생성형 AI 기초',
  },
  {
    value: 'travel_outdoor',
    label: '여행/아웃도어',
    description: '여행지, 캠핑과 야외 경험을 생생하게 전하는 콘텐츠',
    topicPlaceholder: '예: 대중교통으로 떠나는 당일치기 여행',
  },
  {
    value: 'kids',
    label: '키즈',
    description: '아이와 가족이 함께 보고 즐길 수 있는 콘텐츠',
    topicPlaceholder: '예: 아이와 함께하는 쉬운 과학 놀이',
  },
  {
    value: 'custom',
    label: '직접 입력',
    description: '목록에 없는 분야를 직접 적어 추천에 반영합니다.',
    topicPlaceholder: '예: 선택한 분야에서 다루고 싶은 구체적인 주제',
  },
]

export const CONTENT_IDEA_RECOMMENDATION_LIMITS = {
  topicMin: 2,
  topic: 200,
  targetAudience: 500,
  contentFormat: 100,
  keywords: 10,
  keyword: 50,
  referenceContext: 1500,
  customGenre: 60,
  recommendationCountMin: 1,
  recommendationCountMax: 8,
}

export const EMPTY_CONTENT_IDEA_FILTERS = {
  search: '',
  status: '',
  platform: '',
  priority: '',
  source: '',
  sort: 'created_desc',
}

export const CONTENT_IDEA_SORT_OPTIONS = [
  { value: 'created_desc', label: '최근 등록순' },
  { value: 'updated_desc', label: '최근 수정순' },
  { value: 'priority_desc', label: '우선순위 높은순' },
  { value: 'priority_asc', label: '우선순위 낮은순' },
  { value: 'title_asc', label: '제목 가나다순' },
  { value: 'title_desc', label: '제목 역순' },
  { value: 'created_asc', label: '오래된 등록순' },
  { value: 'updated_asc', label: '오래된 수정순' },
]

export const CONTENT_IDEA_LIMITS = {
  title: 200,
  description: 5000,
  targetAudience: 500,
  contentFormat: 100,
  tags: 20,
  tag: 50,
}
